File: robot.rb
class Robot
  VALID_X, VALID_Y = 0..4, 0..4
  VALID_F = %w(north east south west)

  def initialize
    @x, @y, @f = 0, 0, nil
  end

  def place(x, y, f)
    x, y = x.to_i, y.to_i
    @x, @y, @f = x, y, f if VALID_X.include?(x) && VALID_Y.include?(y) && VALID_F.include?(f)
  end

  def move
    return unless @f

    next_x = case @f
    when 'east' then @x + 1
    when 'west' then @x - 1
    end

    next_y = case @f
    when 'north' then @y + 1
    when 'south' then @y - 1
    end

    @x = next_x if VALID_X.include?(next_x)
    @y = next_y if VALID_Y.include?(next_y)
  end

  def rotate(direction)
    return unless current_index = VALID_F.index(@f)

    case direction
    when 'left'
      @f = VALID_F.fetch(current_index - 1)
    when 'right'
      @f = VALID_F.fetch(current_index + 1, VALID_F.first)
    end
  end

  def report
    "#{@x},#{@y},#{@f}"
  end
end

File: javascript/robot.js
const XY = [0, 1, 2, 3, 4];
const F = ["north", "east", "south", "west"];

class Robot {
  place(a) {
    if (a.length === 3 && [XY, XY, F].every((v, i) => v.includes(a[i])))
      (this.x = a[0]), (this.y = a[1]), (this.f = a[2]);
  }

  move() {
    const axis = ["north", "south"].includes(this.f) ? "y" : "x";
    const nextVal = this[axis] + (["north", "east"].includes(this.f) ? 1 : -1);
    if (XY.includes(nextVal)) this[axis] = nextVal;
  }

  rotate(dir) {
    const nextIndex = F.indexOf(this.f) + (dir === "left" ? -1 : 1);
    if (this.f) this.f = F[nextIndex] || F[dir === "left" ? 3 : 0];
  }
}

module.exports = Robot;
